fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -611,24 +751,22 @@ Future<T> kmsRequestImpl(
     std::function<T(Reference<RESTKmsConnectorCtx>, Reference<HTTP::IncomingResponse>)> parseFunc) {
 	state UID requestID = deterministicRandom()->randomUniqueID();
 
-	// Follow 2-phase scheme:
-	// Phase-1: Attempt to do request reaching out to cached KmsUrls in the order of
-	//          past success requests success counts.
-	// Phase-2: For some reason if none of the cached KmsUrls worked, re-discover the KmsUrls and
-	//          repeat phase-1.
+	// Follow multi-phase approach:
+	// Step-1: Enumerate KmsUrlStore cached URLs in the defined order of preference, if URL fails with an acceptable
+	// error (time-out or connection-failed), then continue enumeration. Otherwise, bubble up the error.
+	// Step-2: Refresh KmsUlrStore cached URLs by re-discovering KMS URLs and loop Step-1
 
-	state int pass = 1;
-	for (; pass <= 2; pass++) {
-		state std::stack<std::shared_ptr<KmsUrlCtx>> tempStack;
-
-		// Iterate over Kms URLs
-		while (!ctx->kmsUrlHeap.empty()) {
-			state std::shared_ptr<KmsUrlCtx> curUrl = ctx->kmsUrlHeap.top();
-			ctx->kmsUrlHeap.pop();
-			tempStack.push(curUrl);
+	state int pass = 0;
+	state KmsUrlCtx<KmsUrlPenaltyParams>* urlCtx;
+	loop {
+		state int idx = 0;
+		state int64_t start = now();
 
+		pass++;
+		while (idx < ctx->kmsUrlStore.kmsUrls.size()) {
+			urlCtx = &ctx->kmsUrlStore.kmsUrls[idx++];
```

Comment:
i think this algorithm has a pathology, i'm not sure if it is intended.
If one url gets any sort of penalty, and is now scored lower than the others, we will never send it another request, until another url also gets an error and gets pushed down in the list.
For this type of algorithm to work, we have to make sure that we still sometimes balance requests to the non-first url or periodically do canary requests to the other urls to make sure that if they're back up, we reset their penalty.

Reply:
Talked offline with Josh. 
Idea is to implement a scheme to load-balance requests to KMS avoiding hot-spotting. For now will keep the code same and run more scaling experiments, with the newer datasets we might get more insight into what optimizations could be done.